ruff_cli/src/printer.rs

```diff
@@ -336,6 +343,56 @@ impl<'a> Printer<'a> {
         Ok(())
     }
 
+    pub fn write_statistics(&self, diagnostics: &Diagnostics) -> Result<()> {
+        let mut violations = diagnostics
+            .messages
+            .iter()
+            .map(|message| message.kind.rule())
+            .collect::<Vec<_>>();
+        violations.sort();
+        violations.dedup();
+        let statistics = violations.iter().map(|rule| ExpandedStatistics {
+            code: rule.code(),
+            count: diagnostics
+                .messages
+                .iter()
+                .filter(|message| message.kind.rule() == *rule)
+                .count(),
+            message: diagnostics
+                .messages
+                .iter()
+                .find(|message| message.kind.rule() == *rule)
+                .map(|message| message.kind.body())
+                .unwrap(),
+        });
+
+        let mut stdout = BufWriter::new(io::stdout().lock());
+        match self.format {
+            SerializationFormat::Json => {
+                writeln!(
+                    stdout,
+                    "{}",
+                    serde_json::to_string_pretty(&statistics.collect::<Vec<_>>())?
+                )?;
+            }
+            SerializationFormat::Junit => {}
+            SerializationFormat::Text => {
+                for msg in statistics {
+                    writeln!(stdout, "{:<5} {} {}", msg.count, msg.code, msg.message)?;
+                }
+                return Ok(());
+            }
+            SerializationFormat::Grouped => {}
+            SerializationFormat::Github => {}
+            SerializationFormat::Gitlab => {}
+            SerializationFormat::Pylint => {}
```

Comment:
Shouldn't these rather `anyhow::bail!()`?

Reply:
added that. can we handle that on argparsing level somehow as well?

Reply:
For a clean solution where we have a different enum we'd have to turn `statistics` into a subcommand ... I don't think we want to make it a top-level subcommand and a subsubcommand for the `checker` subcommand might be too confusing for the user.

Otherwise we can probably somehow hook into clap's conflict system but we'd still have the same enum so I don't really see the point of that ... it's better if this is defined in the printer.

Reply:
the subcommand thing has been developed in the last days, right?
so the regular `ruff` command should be `ruff check` in the future?
i consider the additional output of statistics still as a option then.
`ruff statistics` or `ruff check statistics` doesn't fit for my feeling.

Reply:
Yes I agree since it makes sense to combine `--statistics` with the other options of the (implicit) `check` command it does make sense to keep it as an option :)